Update focus mode UI

connorff opened the pull request:
## Milestones:

- [x] Add focus mode status bar item
- [x] Handle updates to a file when focus mode is active

## Overview

### Status bar item
The crux of the PR is the addition of a vscode `StatusBarItem` which displays information about Flowistry and focus mode. The item displays a "loading" status while child processes are running (flowistry subcommands, rustup commands, etc) which indicates to the user that Flowistry is doing _something_. The item is also heavily used by focus mode to indicate the state of the mode and allow for easy toggling and error viewing.

One of my concerns about the PR is the status bar item's scope (defined globally in `extension.ts`). It needs to be globally scoped so it's accessible by `exec_notify`, the function responsible for spawning processes. It's not _terrible_ as-is but I haven't come up with a solution that doesn't require drilling the item through every function call to ensure it's accessible by `exec_notify` as a parameter.

### `FocusMode` class
This PR refactors the functions from `focus.ts` into a `FocusMode` class. The class still uses the initialize -> render flow, but adds a couple things:

- keeps track of the status of focus mode (active, inactive, error, etc) which updates the status bar item color and icon
- adds event callbacks for document edits and saves (and toggles rendering decorations accordingly)
- exposes command functions (focus, mark) as methods

### `CallFlowistry` error handling
The PR also adds two `Error` classes which implement a `FlowistryError` interface:

- `FlowistryBuildError`: running a `flowistry` subcommand through `exec_notify` raises an error
- `FlowistryRuntimeError`: the parsed output of the `flowistry` subcommand is an "Err" variant

Both classes implement a `show` method which displays the error in the same way as previously (notification or document popup). I guess both errors are technically at runtime so if you prefer different names that makes sense.

## Demo

https://user-images.githubusercontent.com/40175891/153699266-9c957944-d96e-4f28-894c-6b979ca82636.mp4

connorff commented on ide/src/extension.ts:
```diff
@@ -1,13 +1,16 @@
 
 import "./app.scss";
 
+export let focus_status_bar_item: vscode.StatusBarItem;
```
Mutable global variables feels really messy - brainstorming better ways to manage state.

willcrichton replied:
Determination after conversation was to use a class to encapsulate global state.

connorff replied:
Not resolving yet since this can't be encapsulated in `FocusMode` only. Same with `tdcp` also defined in `extension.ts`.

connorff commented on ide/src/extension.ts:
```diff
@@ -23,6 +26,7 @@ export async function activate(context: vscode.ExtensionContext) {
       ["focus_unmark", focus_unmark],
       ["focus_select", focus_select],
       ["decompose", decompose],
+      ["last_error", last_error.bind(context)],
```
A click on a [`StatusBarItem`](https://code.visualstudio.com/api/references/vscode-api#StatusBarItem) is mapped to a vscode command so we need a command which eventually opens the error document. Usually, clicking on the status bar item will toggle focus mode (by calling `flowistry.focus`), but the command switches to `flowistry.last_error` when in an error state.

An alternative is to display the error document when there's both an error in the program and the user is deactivating focus mode. This has the benefit of not creating a new command just for showing the error, but also leads to issues when the user just wants to exit focus mode while the program doesn't compile.

connorff commented on ide/src/focus.ts:
```diff
@@ -26,8 +28,9 @@ interface FocusState {
+let doc_save_callback: vscode.Disposable;
 
-let initialize = async (call_flowistry: CallFlowistry) => {
+let initialize = async (call_flowistry: CallFlowistry, hide_error?: boolean) => {
```
The `hide_error` flag prevents the compiler output from displaying each time the user saves the document while in focus mode.

connorff commented on ide/src/focus.ts:
```diff
@@ -147,8 +154,13 @@ export let focus = async (call_flowistry: CallFlowistry) => {
     await initialize(call_flowistry);
     render();
+
+    doc_save_callback = vscode.workspace.onDidSaveTextDocument(() => initialize(call_flowistry, true));
```
This will probably be changed when I work on the UI during/between edits, but for now I just reinitialize focus mode after each document save. The only different between the initialization from user input (win + r, win + a) and initialization from document saves is that the latter won't display compiler output if the program contains errors.

I've also only tested this on simple functions in a small crate so I haven't run into performance issues which will happen when reinitializing after every save.

connorff commented on ide/src/setup.ts:
```diff
@@ -210,10 +195,14 @@ export async function setup(
+        await vscode.window.showTextDocument(doc, vscode.ViewColumn.Beside);
+      }
+      
+      context.workspaceState.update('err_log', e);
```
This seems like the way to go for state management that can be stringifyed.

willcrichton commented on ide/src/setup.ts:
```diff
@@ -177,27 +177,12 @@ export async function setup(
   );
 
-  return async <T>(args: string) => {
+  return async <T>(args: string, hide_error?: boolean) => {
```
Rather than add this parameter, I think the cleaner approach is to move the error rendering out of `CallFlowistry` and into the extensions. One possibility is to have this function throw two types of exceptions to distinguish build errors vs. Flowistry errors.

willcrichton commented on ide/src/setup.ts:
```diff
@@ -210,10 +195,14 @@ export async function setup(
-      let doc = await vscode.workspace.openTextDocument(tdcp.uri);
-      await vscode.window.showTextDocument(doc, vscode.ViewColumn.Beside);
+      if (!hide_error) {
+        tdcp.contents = e.toString();
```
This seems like it's duplicating code with `last_error`.

willcrichton commented on ide/src/focus_utils.ts:
```diff
@@ -0,0 +1,45 @@
+import vscode from 'vscode';
+
+export enum FocusStatusBarState {
```
You can just use `FocusStatusBarState = "active" | "inactive" | "error"`. That is, sadly, the more idiomatic Typescript.

connorff commented on ide/src/extension.ts:
```diff
@@ -39,6 +43,11 @@ export async function activate(context: vscode.ExtensionContext) {
       context.subscriptions.push(disposable);
     });
+
+    focus_status_bar_item = vscode.window.createStatusBarItem();
```
Since the scope of the status bar (as I understand) is to also indicate when Flowistry is computing/executing something, I needed to define this outside just the scope of the `FocusMode` class (same as global state comment).

connorff commented on ide/src/decompose.ts:
```diff
@@ -45,59 +46,53 @@ export let decompose = async (call_flowistry: CallFlowistry) => {
-      return;
-    }
+  let cmd = `decompose ${doc.fileName} ${doc.offsetAt(selection.anchor)}`;
+  let decomp = await call_flowistry<Decomposition>(cmd);
+  if (is_flowistry_error(decomp)) {
+    return decomp.show();
+  }
```
This is applicable for all `call_flowistry` calls, but the idea is we can remove the `try { ... }` block by returning `T | FlowistryError`. The error is caught and displayed depending on its implementation of `show`.

My only worry about this is catching errors unrelated to flowistry (maybe raised by the extension?). It is also a matter of style preference so if you have thoughts I'd be very open.

connorff commented on ide/src/extension.ts:
```diff
@@ -1,28 +1,48 @@
 
 import "./app.scss";
 
+export let flowistry_status_bar_item: vscode.StatusBarItem;
+export const tdcp = new FlowistryErrorDocument();
```
Still might need some global state 😔

The text document provider can only be defined once to use the same uri scheme and the status bar item is imported/modified in a couple distinct functions. Mentioned this in the PR description, but this could be solved by propagating/drilling these into each function they're used (although this is arguably worse than making them globally accessible).

connorff commented on ide/src/extension.ts:
```diff
@@ -1,28 +1,48 @@
-      ["focus_mark", focus_mark],
-      ["focus_unmark", focus_unmark],
-      ["focus_select", focus_select],
+      ["focus", focus.focus.bind(focus)],
+      ["focus_mark", focus.focus_mark.bind(focus)],
+      ["focus_unmark", focus.focus_unmark.bind(focus)],
+      ["focus_select", focus.focus_select.bind(focus)],
```
Each `focus.____` method loses its `this` context when registered as a command ([happens when passed to a HOF](https://eliux.github.io/javascript/common-errors/why-this-gets-undefined-inside-methods-in-javascript/)). Need to explicitly bind `focus` to access `this`.

willcrichton replied:
If you define the class methods as arrow functions, you won't need to call `.bind(...)`.

willcrichton commented on ide/src/error_types.ts:
```diff
@@ -0,0 +1,38 @@
+import vscode from "vscode";
+import { log, show_error } from "./vsc_utils";
+import { tdcp } from "./extension";
+
```
I think a simpler design might be:

```ts
interface BuildError { type: "build-error", error: string }
interface AnalysisError { type: "analysis-error", error: string }
interface FlowistryOutput<T> { type: "output", value: T {
type FlowistryResult<T> = FlowistryOutput<T> | BuildError | AnalysisError

let show = (error: BuildError | AnalysisError) => {
  // ...
}

let is_ok = (res: FlowistryResult<T>): res is FlowistryOutput<T> => res.type == "output";
```

Then when you're using it:

```ts
let result = call_flowistry(...);
if (!is_ok(result)) {
  show(result);
  return;
}

process(result);
```

A few reasons:
* This avoids the `result as T` casting that's necessary otherwise.
* This reduces the boilerplate with error interfaces and class definitions. 
* We don't need errors that extend `Error` since we're not throwing exceptions.

willcrichton commented on ide/src/extension.ts:
```diff
@@ -1,28 +1,48 @@
-      ["focus_mark", focus_mark],
-      ["focus_unmark", focus_unmark],
-      ["focus_select", focus_select],
+      ["focus", focus.focus.bind(focus)],
```
Let's move this sub-list to a `FocusMode` method, like `focus.commands()` or something.

willcrichton commented on ide/src/focus.ts:
```diff
@@ -25,130 +27,224 @@ interface FocusState {
 
-let state: FocusState | null = null;
+export class FocusMode {
+  status = "inactive";
```
should have a type like `status: "inactive" | "loading" | ...`

willcrichton commented on ide/src/focus.ts:
```diff
@@ -25,130 +27,224 @@ interface FocusState {
   disposable: vscode.Disposable;
 }
 
```
This class is now sufficiently complicated / stateful that it can use a doc-comment explaining it's lifecycle. Like it would be useful to say what the states are, and how the user's interactions change the states and call specific functions.

connorff commented on ide/src/focus.ts:
```diff
@@ -25,130 +27,224 @@ interface FocusState {
-    }
-  } else {
-    clear_ranges(active_editor);
+  private setStatus(status: FocusStatus) {
```
React `setState` took the wheel... 😔

connorff commented on ide/src/focus.ts:
```diff
@@ -25,130 +27,224 @@ interface FocusState {
+  async initialize(hide_error: boolean = false) {
+    this.add_watchers();
+
     let active_editor = vscode.window.activeTextEditor;
```
Just to clarify, `initialize` should return a `FlowistryResult` (from your suggestion) and have its callers handle the logic for error display?

willcrichton replied:
No I'm just referring to checking if there is an active text editor.

connorff replied:
It seems each call of `initialize` is guarded by a call to `this.is_active()`. To ensure focus is only active when there's an active editor, I think I'd need to add a callback for `onDidChangeActiveTextEditor` to `add_watchers`.

Does that sound fine?

willcrichton replied:
Actually don't worry about this for now. Once this PR is in I'm going to change how we deal with multiple documents.

connorff commented:
All suggestions should now be addressed. I also made a few changes for clarity/cleanup:

- Added `private` to all `Focus` methods that aren't used by class consumers to clean up API
- Moved the `onDidChangeTextEditorSelection` callback to `add_watchers` to be more consistent with the other onDid[x] handlers
- Removed the `Focus.should_render` method - rendering shouldn't be allowed with the `loading` status so we can just check that focus mode is active to decide whether or not to render

The PR also isn't 50% merge commits which is nice :)

willcrichton commented:
Looks great! I might make a few small changes but I'll do that myself. Thanks for all the hard work :-)
